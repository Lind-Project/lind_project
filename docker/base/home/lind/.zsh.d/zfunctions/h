## Fix zsh annoying history behavior
 if [[ -z "$*" ]]; then history; else egrep "$*" <(fc -l 1); fi
