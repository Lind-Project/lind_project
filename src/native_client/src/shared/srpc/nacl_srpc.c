/*
 * Copyright (c) 2011 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/*
 * NaCl service library.  a primitive rpc library
 */

#include <stdarg.h>
#include <stdlib.h>
#include <string.h>

#ifndef __native_client__
#include "native_client/src/include/portability.h"
#include "native_client/src/trusted/desc/nacl_desc_imc.h"
#endif  /* __native_client__ */
#include "native_client/src/shared/srpc/nacl_srpc.h"
#include "native_client/src/shared/platform/nacl_check.h"
#include "native_client/src/shared/srpc/nacl_srpc_internal.h"
#include "native_client/src/shared/srpc/nacl_srpc_message.h"

/*
 * Service discovery is used to build an interface description that
 * is searched for rpc dispatches.
 */
static NaClSrpcService* BuildClientService(NaClSrpcChannel* channel) {
  NaClSrpcArg*         ins[] = { NULL };
  NaClSrpcArg          out_carray;
  NaClSrpcArg*         outs[2];
  NaClSrpcService*     tmp_service = NULL;
  NaClSrpcService*     service = NULL;
  NaClSrpcHandlerDesc  basic_services[] = { { NULL, NULL } };
  NaClSrpcError        service_discovery_result;
  NaClSrpcService*     returned_service = NULL;

  /*
   * Service discovery requires a temporary setting of channel->client.
   * To ensure that this does not clobber a previous setting, enforce
   * a precondition that it is NULL beforehand.
   */
  CHECK(NULL == channel->client);
  /* Set up the output parameters for service discovery. */
  NaClSrpcArgCtor(&out_carray);
  outs[0] = &out_carray;
  outs[1] = NULL;
  out_carray.tag = NACL_SRPC_ARG_TYPE_CHAR_ARRAY;
  out_carray.u.count = NACL_SRPC_MAX_SERVICE_DISCOVERY_CHARS;
  /*
   * Add one guaranteed NUL byte to make sure the string will be NUL-terminated.
   */
  out_carray.arrays.carr = calloc(NACL_SRPC_MAX_SERVICE_DISCOVERY_CHARS + 1, 1);
  if (NULL == out_carray.arrays.carr) {
    NaClSrpcLog(NACL_SRPC_LOG_ERROR,
                "BuildClientService(channel=%p):"
                " service discovery malloc failed\n",
                (void*) channel);
    goto cleanup;
  }
  /* Set up the basic service descriptor to make the service discovery call. */
  tmp_service = (NaClSrpcService*) malloc(sizeof(*tmp_service));
  if (NULL == tmp_service) {
    NaClSrpcLog(NACL_SRPC_LOG_ERROR,
                "BuildClientService(channel=%p):"
                " temporary service malloc failed\n",
                (void*) channel);
    goto cleanup;
  }
  if (!NaClSrpcServiceHandlerCtor(tmp_service, basic_services)) {
    NaClSrpcLog(NACL_SRPC_LOG_ERROR,
                "BuildClientService(channel=%p):"
                " NaClSrpcServiceHandlerCtor failed\n",
                (void*) channel);
    free(tmp_service);
    tmp_service = NULL;
    goto cleanup;
  }
  /* Temporarily assign channel->client to allow Invoke. */
  channel->client = tmp_service;
  /* Invoke service discovery, getting description string */
  service_discovery_result = NaClSrpcInvokeV(channel, 0, ins, outs);
  /* Disconnect the temporary service. */
  channel->client = NULL;
  if (NACL_SRPC_RESULT_OK != service_discovery_result) {
    NaClSrpcLog(NACL_SRPC_LOG_ERROR,
                "BuildClientService(channel=%p):"
                " service discovery invoke failed\n",
                (void*) channel);
    goto cleanup;
  }
  /* Allocate the real service. */
  service = (NaClSrpcService*) malloc(sizeof(*service));
  if (NULL == service) {
    NaClSrpcLog(NACL_SRPC_LOG_ERROR,
                "BuildClientService(channel=%p):"
                " service discovery malloc failed\n",
                (void*) channel);
    goto cleanup;
  }
  /* Build the real service from the resulting string. */
  if (!NaClSrpcServiceStringCtor(service, out_carray.arrays.carr)) {
    NaClSrpcLog(NACL_SRPC_LOG_ERROR,
                "BuildClientService(channel=%p):"
                " NaClSrpcServiceStringCtor failed\n",
                (void*) channel);
    goto cleanup;
  }
  /* Caller takes ownership of service. */
  returned_service = service;
  service = NULL;
 cleanup:
  free(service);
  NaClSrpcServiceDtor(tmp_service);
  free(tmp_service);
  free(out_carray.arrays.carr);
  return returned_service;
}

/*
 * The constructors and destructor.
 * Clients and Servers are both NaClSrpcChannels, but they are constructed
 * by different methods reflecting how they get their table of rpc methods.
 */

void NaClSrpcChannelInitialize(NaClSrpcChannel* channel) {
  channel->message_channel = NULL;
  channel->server = NULL;
  channel->client = NULL;
  channel->server_instance_data = NULL;
}

static int NaClSrpcChannelCtorHelper(NaClSrpcChannel* channel,
                                     NaClSrpcImcDescType handle) {
  NaClSrpcLog(1,
              "NaClSrpcChannelCtorHelper(channel=%p, handle=%p)\n",
              (void*) channel,
              (void*) handle);
  /* Initialize the channel to allow errors to be handled. */
  NaClSrpcChannelInitialize(channel);
  /* Create the message fragmentation layer. */
  channel->message_channel = NaClSrpcMessageChannelNew(handle);
  if (NULL == channel->message_channel) {
    NaClSrpcLog(NACL_SRPC_LOG_ERROR,
                "NaClSrpcChannelCtorHelper(channel=%p):"
                " NaClSrpcMessageChannelNew failed\n",
                (void*) channel);
    return 0;
  }
  return 1;
}

static void NaClSrpcChannelDtorHelper(NaClSrpcChannel* channel) {
  NaClSrpcLog(1, "NaClSrpcChannelDtorHelper(channel=%p)\n", (void*) channel);
  channel->server_instance_data = NULL;
  NaClSrpcMessageChannelDelete(channel->message_channel);
  channel->message_channel = NULL;
}

int NaClSrpcClientCtor(NaClSrpcChannel* channel, NaClSrpcImcDescType handle) {
  int return_value                = 0;
  NaClSrpcService* client_service = NULL;
  NaClSrpcLog(1,
              "NaClSrpcClientCtor(channel=%p, handle=%p)\n",
              (void*) channel,
              (void*) handle);
  if (!NaClSrpcChannelCtorHelper(channel, handle)) {
    goto cleanup;
  }
  /* Do service discovery to speed method invocation. */
  client_service = BuildClientService(channel);
  if (NULL == client_service) {
    NaClSrpcLog(NACL_SRPC_LOG_ERROR,
                "NaClSrpcClientCtor(channel=%p):"
                " BuildClientService failed\n",
                (void*) channel);
    NaClSrpcChannelDtorHelper(channel);
    goto cleanup;
  }
  /* Channel takes ownership of client_service. */
  channel->client = client_service;
  client_service = NULL;
  return_value = 1;
 cleanup:
  NaClSrpcServiceDtor(client_service);
  return return_value;
}

int NaClSrpcServerCtor(NaClSrpcChannel* channel,
                       NaClSrpcImcDescType handle,
                       NaClSrpcService* service,
                       void* server_instance_data) {
  int return_value = 0;
  NaClSrpcLog(1,
              "NaClSrpcServerCtor(channel=%p, handle=%p,"
              " service=%p, server_instance_data=%p)\n",
              (void*) channel,
              (void*) handle,
              (void*) service,
              (void*) server_instance_data);
  if (!NaClSrpcChannelCtorHelper(channel, handle)) {
    goto cleanup;
  }
  /* Channel takes ownership of service. */
  channel->server = service;
  /* Channel does not take ownership of server_instance_data. */
  channel->server_instance_data = server_instance_data;
  return_value = 1;
 cleanup:
  return return_value;
}

void NaClSrpcDtor(NaClSrpcChannel* channel) {
  NaClSrpcLog(1,
              "NaClSrpcDtor(channel=%p)\n",
              (void*) channel);
  if (NULL == channel) {
    return;
  }
  channel->server_instance_data = NULL;
  NaClSrpcServiceDtor(channel->client);
  free(channel->client);
  channel->client = NULL;
  NaClSrpcServiceDtor(channel->server);
  free(channel->server);
  channel->server = NULL;
  /* Do the common part destruction. */
  NaClSrpcChannelDtorHelper(channel);
}

void NaClSrpcArgCtor(NaClSrpcArg* arg) {
  memset(arg, 0, sizeof *arg);
}

/*
 * A standalone SRPC server is not a subprocess of the browser or
 * sel_universal.  As this is a mode used for testing, the parent environment
 * must set the following variable to indicate that fact.
 */
const char kSrpcStandalone[] = "NACL_SRPC_STANDALONE";

int NaClSrpcIsStandalone(void) {
  return (NULL != getenv(kSrpcStandalone));
}
