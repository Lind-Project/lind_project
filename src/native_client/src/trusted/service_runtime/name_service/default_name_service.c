/*
 * Copyright (c) 2011 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

#include "native_client/src/trusted/service_runtime/name_service/default_name_service.h"

#include "native_client/src/shared/platform/nacl_log.h"
#include "native_client/src/trusted/desc/nacl_desc_rng.h"
#include "native_client/src/trusted/manifest_name_service_proxy/manifest_proxy.h"
#include "native_client/src/trusted/service_runtime/include/sys/fcntl.h"
#include "native_client/src/trusted/service_runtime/sel_ldr_thread_interface.h"

int NaClDefaultNameServiceInit(struct NaClNameService *ns) {
  /*
   * Create an CSPRNG and enter it into the name server.
   */
  struct NaClDescRng  *rng = NULL;

  rng = (struct NaClDescRng *) malloc(sizeof *rng);
  if (NULL == rng) {
    goto malloc_failed;
  }
  if (!NaClDescRngCtor(rng)) {
    goto rng_ctor_failed;
  }

  /*
   * It may appear desirable to insert a factory for rng, so there can
   * be per-thread secure rng access.  However, note that the only way
   * we "transfer" a RNG is to create a new (but indistinguishable)
   * RNG at the recipient, so each lookup results in a new generator
   * anyway.
   */
  (*NACL_VTBL(NaClNameService, ns)->
   CreateDescEntry)(ns,
                    "SecureRandom", NACL_ABI_O_RDWR,
                    (struct NaClDesc *) rng);
  rng = NULL;

  return 1;

 rng_ctor_failed:
  free(rng);
 malloc_failed:
  return 0;
}
