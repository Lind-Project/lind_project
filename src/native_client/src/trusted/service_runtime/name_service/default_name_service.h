/*
 * Copyright 2011 The Native Client Authors.  All rights reserved.
 * Use of this source code is governed by a BSD-style license that can
 * be found in the LICENSE file.
 */

#ifndef NATIVE_CLIENT_SRC_TRUSTED_SERVICE_RUNTIME_NAME_SERVICE_DEFAULT_NAME_SERVICE_H_
#define NATIVE_CLIENT_SRC_TRUSTED_SERVICE_RUNTIME_NAME_SERVICE_DEFAULT_NAME_SERVICE_H_

#include "native_client/src/include/nacl_base.h"
#include "native_client/src/include/portability.h"

#include "native_client/src/trusted/desc/nacl_desc_base.h"

#include "native_client/src/trusted/service_runtime/name_service/name_service.h"

int NaClDefaultNameServiceInit(struct NaClNameService *ns);

#endif
