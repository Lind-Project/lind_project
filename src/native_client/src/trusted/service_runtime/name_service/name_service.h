/*
 * Copyright (c) 2011 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

#ifndef NATIVE_CLIENT_SRC_TRUSTED_SERVICE_RUNTIME_NAME_SERVICE_NAME_SERVICE_H_
#define NATIVE_CLIENT_SRC_TRUSTED_SERVICE_RUNTIME_NAME_SERVICE_NAME_SERVICE_H_

#include "native_client/src/include/nacl_base.h"
#include "native_client/src/include/portability.h"

/* Get NACL_NAME_SERVICE_CONNECTION_MAX. */
#include "native_client/src/public/name_service.h"

#include "native_client/src/trusted/desc/nacl_desc_base.h"

#include "native_client/src/trusted/simple_service/nacl_simple_service.h"
#include "native_client/src/trusted/simple_service/nacl_simple_ltd_service.h"

#include "native_client/src/trusted/threading/nacl_thread_interface.h"

EXTERN_C_BEGIN

/*
 * NaClStartNameService should only be called once, and also serves
 * as a module initializer.
 *
 * This function creates a bound socket / socket address pair, and
 * spawns a thread that accepts connections on the bound socket and
 * services requests on the resultant connections.  A maximum number
 * of connections is defined, since well-behaving programs should not
 * need more than one (absent excessive opaque layering/abstractions)
 * name service lookup connection.
 *
 * The socket address NaClDescConnCap is returned, to be entered into
 * the NaClApp's open file table at a well-known location.
 */

struct NaClNameServiceEntry;  /* fwd */

struct NaClNameService {
  struct NaClSimpleLtdService base NACL_IS_REFCOUNT_SUBCLASS;

  struct NaClMutex            mu;
  /*
   * |mu| protects the service entries hanging off of |head|.
   */
  struct NaClNameServiceEntry *head;
};

int NaClNameServiceCtor(struct NaClNameService      *self,
                        NaClThreadIfFactoryFunction thread_factory_fn,
                        void                        *thread_factory_data);

int NaClNameServiceCreateDescEntry(
    struct NaClNameService  *self,
    char const              *name,
    int                     mode,
    struct NaClDesc         *new_desc);  /* takes ownership of ref */

typedef int (*NaClNameServiceFactoryFn_t)(void             *factory_state,
                                          char const       *name,
                                          int              flags,
                                          struct NaClDesc  **out);

int NaClNameServiceCreateFactoryEntry(
    struct NaClNameService      *self,
    char const                  *name,
    /*
     * If a |name| corresponding to a factory is overwritten,
     * references to the factory and its state is just dropped/lost
     * after invoking the factory with a NULL for the |out| argument;
     * the factory function should decrement reference counts as
     * appropriate and clean up when this occurs.
     */
    NaClNameServiceFactoryFn_t  factory_fn,
    void                        *factory_state);

int NaClNameServiceResolveName(
    struct NaClNameService  *self,
    char const              *name,
    int                     flags,
    struct NaClDesc         **out);

int NaClNameServiceDeleteName(struct NaClNameService  *nnsp,
                              char const              *name);

size_t NaClNameServiceEnumerate(struct NaClNameService  *nnsp,
                                char                    *dest,
                                size_t                  nbytes);

struct NaClNameServiceVtbl {
  struct NaClSimpleServiceVtbl  vbase;
  /*
   * The following functions return 0 for success, and non-zero to
   * indicate the reason for failure.  The set of possible failures is
   * in src/public/name_service.h.
   *
   * They don't really have to be virtual, but this makes it easier to
   * subclass and modify later, should the need occur.
   */
  int                           (*CreateDescEntry)(
      struct NaClNameService  *self,
      char const              *name,
      int                     mode,
      struct NaClDesc         *new_desc);  /* takes ownership of ref */
  int                           (*CreateFactoryEntry)(
      struct NaClNameService      *self,
      char const                  *name,
      NaClNameServiceFactoryFn_t  factory_fn,
      void                        *factory_state);
  int                           (*ResolveName)(
      struct NaClNameService  *self,
      char const              *name,
      int                     flags,
      struct NaClDesc         **out);
  int                           (*DeleteName)(
      struct NaClNameService  *self,
      char const              *name);

  /*
   * Enumerate always succeeds and returns the number of bytes
   * written.  To see if the buffer was large enough, check that the
   * number of bytes written is less than the supplied buffer size.
   * NB: if the number of bytes is equal to the buffer size, even if
   * the last byte is a NUL we cannot know that there aren't more
   * entries.  (A partially written entry is only one possibility; it
   * could happen that the initial entries exactly fit the buffer, but
   * there are more.)
   */
  size_t                        (*Enumerate)(
      struct NaClNameService  *self,
      char                    *dest,
      size_t                  nbytes);
};

extern struct NaClNameServiceVtbl kNaClNameServiceVtbl;

void NaClNameServiceLaunch(struct NaClNameService *self);

EXTERN_C_END

#endif
/* NATIVE_CLIENT_SRC_TRUSTED_SERVICE_RUNTIME_NAME_SERVICE_NAME_SERVICE_H_ */
