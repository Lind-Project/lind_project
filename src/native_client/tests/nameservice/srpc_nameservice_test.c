/*
 * Copyright (c) 2012 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/*
 * SRPC-based test for simple rpc based access to name services.
 */

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <inttypes.h>
#include <sys/fcntl.h>
#include <string.h>
#include <unistd.h>

#include "native_client/src/public/imc_syscalls.h"
#include "native_client/src/public/name_service.h"
#include "native_client/src/shared/srpc/nacl_srpc.h"

#define RNG_OUTPUT_BYTES  1024

#define BYTES_PER_LINE    32
#define BYTE_SPACING      4

struct StringBuffer {
  size_t  nbytes;
  size_t  insert;
  char    *buffer;
};

NaClSrpcChannel ns_channel;

void StringBufferCtor(struct StringBuffer *sb) {
  sb->nbytes = 1024;
  sb->insert = 0;
  sb->buffer = malloc(sb->nbytes);
  if (NULL == sb->buffer) {
    perror("StringBufferInit::malloc");
    abort();
  }
  sb->buffer[0] = '\0';
}

void StringBufferDiscardOutput(struct StringBuffer *sb) {
  sb->insert = 0;
  sb->buffer[0] = '\0';
}

void StringBufferDtor(struct StringBuffer *sb) {
  sb->nbytes = 0;
  sb->insert = 0;
  free(sb->buffer);
  sb->buffer = NULL;
}

void StringBufferPrintf(struct StringBuffer *sb,
                        char const *fmt, ...)
    __attribute__((format(printf, 2, 3)));

void StringBufferPrintf(struct StringBuffer *sb,
                        char const *fmt, ...) {
  size_t space;
  char *insert_pt;
  va_list ap;
  int written = 0;
  char *new_buffer;

  va_start(ap, fmt);
  vfprintf(stderr, fmt, ap);
  va_end(ap);

  for (;;) {
    space = sb->nbytes - sb->insert;
    insert_pt = sb->buffer + sb->insert;
    va_start(ap, fmt);
    written = vsnprintf(insert_pt, space, fmt, ap);
    va_end(ap);
    if (written < space) {
      sb->insert += written;
      break;
    }
    /* insufficient space -- grow the buffer */
    new_buffer = realloc(sb->buffer, 2 * sb->nbytes);
    if (NULL == new_buffer) {
      /* give up */
      fprintf(stderr, "StringBufferPrintf: no memory\n");
      break;
    }
    sb->nbytes *= 2;
    sb->buffer = new_buffer;
  }
}

void dump_output(struct StringBuffer *sb, int d, size_t nbytes) {
  uint8_t *bytes;
  int     got;
  int     copied;
  int     ix;

  bytes = malloc(nbytes);
  if (!bytes) {
    perror("dump_output");
    fprintf(stderr, "No memory\n");
    return;
  }
  /* read output */
  for (got = 0; got < nbytes; got += copied) {
    copied = read(d, bytes + got, nbytes - got);
    if (-1 == copied) {
      perror("dump_output:read");
      fprintf(stderr, "read failure\n");
      break;
    }
    printf("read(%d, ..., %zd) -> %d\n", d, nbytes - got, copied);
  }
  /* hex dump it */
  for (ix = 0; ix < got; ++ix) {
    if (0 == (ix & (BYTES_PER_LINE-1))) {
      StringBufferPrintf(sb, "\n%04x:", ix);
    } else if (0 == (ix & (BYTE_SPACING-1))) {
      StringBufferPrintf(sb, " ");
    }
    StringBufferPrintf(sb, "%02x", bytes[ix]);
  }
  StringBufferPrintf(sb, "\n");

  free(bytes);
}

int EnumerateNames(struct StringBuffer *sb, NaClSrpcChannel *nschan) {
  char      buffer[1024];
  uint32_t  nbytes = sizeof buffer;
  char      *p;
  size_t    name_len;

  if (NACL_SRPC_RESULT_OK != NaClSrpcInvokeBySignature(nschan,
                                                       NACL_NAME_SERVICE_LIST,
                                                       &nbytes, buffer)) {
    return 0;
  }
  StringBufferPrintf(sb, "nbytes = %zu\n", (size_t) nbytes);
  if (nbytes == sizeof buffer) {
    fprintf(stderr, "Insufficent space for namespace enumeration\n");
    return 0;
  }
  for (p = buffer; p - buffer < nbytes; p += name_len) {
    name_len = strlen(p) + 1;
    StringBufferPrintf(sb, "%s\n", p);
  }
  return 1;
}

/*
 * Return name service output
 */
void NameServiceDump(NaClSrpcRpc *rpc,
                     NaClSrpcArg **in_args,
                     NaClSrpcArg **out_args,
                     NaClSrpcClosure *done) {
  struct StringBuffer sb;
  StringBufferCtor(&sb);

  if (EnumerateNames(&sb, &ns_channel)) {
    out_args[0]->arrays.str = strdup(sb.buffer);
    rpc->result = NACL_SRPC_RESULT_OK;
  } else {
    rpc->result = NACL_SRPC_RESULT_APP_ERROR;
  }
  done->Run(done);
  StringBufferDtor(&sb);
}

/*
 * Dump RNG output into a string.
 */
void RngDump(NaClSrpcRpc *rpc,
             NaClSrpcArg **in_args,
             NaClSrpcArg **out_args,
             NaClSrpcClosure *done) {
  struct StringBuffer sb;
  NaClSrpcError rpc_result;
  int status;
  int rng;

  StringBufferCtor(&sb);
  rpc_result = NaClSrpcInvokeBySignature(&ns_channel, NACL_NAME_SERVICE_LOOKUP,
                                         "SecureRandom", O_RDONLY,
                                         &status, &rng);
  assert(NACL_SRPC_RESULT_OK == rpc_result);
  printf("rpc status %d\n", status);
  assert(NACL_NAME_SERVICE_SUCCESS == status);
  printf("rng descriptor %d\n", rng);

  dump_output(&sb, rng, RNG_OUTPUT_BYTES);
  out_args[0]->arrays.str = strdup(sb.buffer);
  rpc->result = NACL_SRPC_RESULT_OK;
  done->Run(done);
  close(rng);
  StringBufferDtor(&sb);
}

void ManifestTest(NaClSrpcRpc *rpc,
                  NaClSrpcArg **in_args,
                  NaClSrpcArg **out_args,
                  NaClSrpcClosure *done) {
  struct StringBuffer     sb;
  NaClSrpcError           rpc_result;
  int                     status;
  int                     manifest;
  int                     manifest_conn;
  struct NaClSrpcChannel  manifest_channel;

  /* just get the descriptor for now */
  StringBufferCtor(&sb);
  if (NACL_SRPC_RESULT_OK !=
      (rpc_result =
       NaClSrpcInvokeBySignature(&ns_channel, NACL_NAME_SERVICE_LOOKUP,
                                 "ManifestNameService", O_RDWR,
                                 &status, &manifest))) {
    StringBufferPrintf(&sb, "nameservice lookup RPC failed (%d)\n", rpc_result);
    goto done;
  }
  StringBufferPrintf(&sb, "Got manifest descriptor %d\n", manifest);
  if (-1 == manifest) {
    fprintf(stderr, "nameservice lookup failed, status %d\n", status);
    goto done;
  }
  /* connect to manifest name server */

  manifest_conn = imc_connect(manifest);
  StringBufferPrintf(&sb, "got manifest connection %d\n", manifest_conn);
  if (-1 == manifest_conn) {
    StringBufferPrintf(&sb, "could not connect\n");
    goto done;
  }
  close(manifest);
  if (!NaClSrpcClientCtor(&manifest_channel, manifest_conn)) {
    StringBufferPrintf(&sb, "could not build srpc client\n");
    goto done;
  }
  StringBufferDiscardOutput(&sb);
  StringBufferPrintf(&sb, "ManifestTest: basic connectivity ok\n");

 done:
  out_args[0]->arrays.str = strdup(sb.buffer);
  rpc->result = NACL_SRPC_RESULT_OK;
  done->Run(done);
  StringBufferDtor(&sb);
}

const struct NaClSrpcHandlerDesc srpc_methods[] = {
  /* Export the methods as taking no arguments and returning a string. */
  { "namedump::s", NameServiceDump },
  { "rngdump::s", RngDump },
  { "manifest_test::s", ManifestTest },
  { NULL, NULL },
};

int main(void) {
  int ns;
  int connected_socket;

  if (!NaClSrpcModuleInit()) {
    return 1;
  }
  ns = -1;
  nacl_nameservice(&ns);
  assert(-1 != ns);

  connected_socket = imc_connect(ns);
  assert(-1 != connected_socket);
  if (!NaClSrpcClientCtor(&ns_channel, connected_socket)) {
    fprintf(stderr, "Srpc client channel ctor failed\n");
    return 1;
  }
  close(ns);

  if (!NaClSrpcAcceptClientConnection(srpc_methods)) {
    return 1;
  }
  NaClSrpcModuleFini();
  return 0;
}
