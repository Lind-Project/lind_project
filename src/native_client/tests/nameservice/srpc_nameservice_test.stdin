rpc namedump * s("")
rpc manifest_test * s("")
