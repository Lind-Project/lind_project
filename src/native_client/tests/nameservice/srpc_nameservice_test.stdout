rpc call initiated namedump::s
rpc call complete namedump::s
output 0:  s("nbytes = 47\nKernelService\nManifestNameService\nSecureRandom\n")

rpc call initiated manifest_test::s
rpc call complete manifest_test::s
output 0:  s("ManifestTest: basic connectivity ok\n")

